packages/core/src/sync-realtime/service.ts

```diff
@@ -8,878 +8,633 @@ import {
+      block: pendingLatestBlock,
+      transactions,
+      logs: pendingLogs,
+    });
 
-    this.common.logger.trace({
+    const matchedLogCountText =
+      pendingLogs.length === 1
+        ? "1 matched log"
+        : `${pendingLogs.length} matched logs`;
+    realtimeSyncService.common.logger.info({
       service: "realtime",
-      msg: `Requesting logs from ${
-        latestLocalBlockNumber + 1
-      } to ${newBlockNumber} (network=${this.network.name})`,
+      msg: `Synced ${matchedLogCountText} from block ${pendingLatestBlockNumber} (network=${realtimeSyncService.network.name})`,
     });
 
-    const _logs = await this._eth_getLogs({
-      fromBlock: numberToHex(latestLocalBlockNumber + 1),
-      toBlock: numberToHex(newBlockNumber),
+    realtimeSyncService.onEvent({
+      type: "checkpoint",
+      chainId: realtimeSyncService.network.chainId,
+      checkpoint: {
+        ...maxCheckpoint,
+        blockTimestamp: pendingLatestBlockTimestamp,
+        chainId: realtimeSyncService.network.chainId,
+        blockNumber: pendingLatestBlockNumber,
+      } as Checkpoint,
     });
-    const logs = sortLogs(_logs);
+  } else {
+    realtimeSyncService.common.logger.info({
+      service: "realtime",
+      msg: `Synced 0 matched logs from block ${pendingLatestBlockNumber} (network=${realtimeSyncService.network.name})`,
+    });
+  }
 
-    const matchedLogs = await this.getMatchedLogs(
-      logs,
-      BigInt(newBlockNumber),
-      true,
+  realtimeSyncService.localChain.push({
+    block: syncBlockToLightBlock(pendingLatestBlock),
+    logs: pendingLogs.map(syncLogToLightLog),
+  });
+
+  realtimeSyncService.common.metrics.ponder_realtime_latest_block_number.set(
+    { network: realtimeSyncService.network.name },
+    pendingLatestBlockNumber,
+  );
+  realtimeSyncService.common.metrics.ponder_realtime_latest_block_timestamp.set(
+    { network: realtimeSyncService.network.name },
+    pendingLatestBlockTimestamp,
+  );
+
+  // Determine if a new range has become finalized by evaluating if the
+  // latest block number is 2 * finalityBlockCount >= finalized block number.
+  // Essentially, there is a range the width of finalityBlockCount that is entirely
+  // finalized.
+  const blockMovesFinality =
+    pendingLatestBlockNumber >=
+    realtimeSyncService.finalizedBlock.number +
+      2 * realtimeSyncService.network.finalityBlockCount;
+  if (blockMovesFinality) {
+    const pendingFinalizedBlock = realtimeSyncService.localChain.find(
+      ({ block }) =>
+        block.number ===
+        pendingLatestBlockNumber -
+          realtimeSyncService.network.finalityBlockCount,
+    )!.block;
+
+    // Validate the local chain by re-requesting logs and checking for inconsistencies
+    // between the local and recently fetched data. This is neccessary because
+    // degraded rpc providers have trouble indexing logs near the tip of the chain.
+    const hasInvalidLogs = await validateLocalBlockchainState(
```

Comment:
Seeing as we're not deeming it critical, I think we might want to wrap this call in a try/catch. I could see that eth_getLogs request failing.

Reply:
Good idea